### ISSUE TITLE
GitIgnore fails to handle brace expansion pattern: {tree,a}.{dot,htm,png,ps,pdf}

### ISSUE THREAD

**Issue by diehuxx** [nielsbasjes/codeowners#171]
<img width="406" alt="Screenshot 2024-11-06 at 9 48 13 AM" src="https://github.com/user-attachments/assets/9965642b-a998-457d-887c-800d011f2bb4">

Hello @nielsbasjes --

I encountered a PatternSyntaxException while using `nl.basjes.gitignore.GitIgnore` with the following gitignore pattern: `{tree,a}.{dot,htm,png,ps,pdf}`

The library attempted to convert this to the regex: `^/?(.*/)?{tree,a}\.{dot,htm,png,ps,pdf}(/|$)`
This resulted in the error: `"Illegal repetition near index 10 ^/?(.*/)?{tree,a}{dot,htm,png,ps,pdf}(/|$)"`

This appears to be a valid gitignore. Is this an edge case that can be fixed?

**Comment by nielsbasjes** [comment]
Thanks for the report.
I'll have to do some testing to see how git treats these expressions and see what I can do to make my library behave the same way.

**Comment by nielsbasjes** [comment]
I checked the git manual on this and the { } you are using is not mentioned.
https://git-scm.com/docs/gitignore

So where did you get this from (documentation with some examples would be great), what should it match and does git actually match what you expect?

**Comment by nielsbasjes** [comment]
The only reference I have been able to find to gitignore and these curly braces is the post which states it is not supported by git.
https://www.reddit.com/r/git/s/3LAC25hLB5

**Comment by nielsbasjes** [comment]
I have done a test:
- Created empty directory and made it a git repo. Added dummy file `foo` to ensure it has atleast 1 commit.

.gitignore 
```
{tree,a}.{dot,htm,png,ps,pdf}
ignore.txt
```

Created the files
- ignore.txt  : To see what happens with a really ignored file.
- Files that should be ignored according to your report
  - a.dot, a.png, tree.dot, tree.pdf, tree.png
- Files that should NOT be ignored according to your report
  - bar.pdf foo.png

Then I do `git status`
```
On branch main
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	.gitignore
	a.dot
	a.png
	bar.pdf
	foo.png
	tree.dot
	tree.pdf
	tree.png
```

The `ignore.txt` shows that any file ignored by git is not shown in this list.

All the files you expect to be ignored are shown, so they are NOT ignored.

So the curly braces expression you provided simply does not work in git as you expected it to. 

This library follows the actually supported gitignore syntax and the error is a mistake at your end.

Note that I have used all the examples and expressions as testcases that are documented on

- https://git-scm.com/docs/gitignore
- https://www.atlassian.com/git/tutorials/saving-changes/gitignore
